rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,697 @@
+
+After a few seconds, the page displays a new message:
+
+    The Database Service has established network connectivity with the your Database!
+
+The user is now allowed to proceed with the remaining steps in the enrollment flow.
+
+> [!NOTE]
+> As of writing, one of the post-deployment steps in the enrollment wizard is a "connection tester", which tests a combination of network connectivity, Teleport RBAC, AWS IAM permissions, and RDS IAM config in the database itself.
+> The problem is that if the connection tester finds a network connectivity issue, then the user would have to go back to redeploy the database agent to fix it.
+> Network health checks can be used to ensure that the deployment's network settings are correct before the user moves on from the deployment step.
+> The other failure modes do not depend on the deployment settings, so it makes sense to keep those tests in a separate subsequent step.
+
+#### User story: web UI resource health indicators
+
+Alice has deployed a Teleport database agent and enrolled a PostgreSQL database with health checks enabled and using the default settings.
+
+Some time later, another user reconfigures the database firewall rules to restrict the allowed inbound traffic IP ranges, but they inadvertently block the Teleport database agent from reaching the database over the network.
+
+Alice logs into the web UI and navigates to the resources page. 
+
+Alice sees a tooltip or warning icon on the database she set up earlier.
+
+Alice clicks on the tooltip/icon to see the message:
+
+    The Teleport database service proxying access to this database cannot reach the database endpoint.
+    <link to health check or network troubleshooting doc>
+
+    Affected Teleport database service:
+    - Hostname: <hostname>
+      UUID: <uuid>
+      Error: "(tcp) failed: Operation timed out"
+
+Among other suggestions, the documentation includes:
+- check that the database is actually listening on `<port>`
+- check that database inbound TCP traffic is allowed from the agent to `<ip : port>`
+- check that agent outbound TCP traffic is allowed to `<ip : port>`
+
+After checking the linked documentation and some investigation Alice diagnoses that the timeout is caused by the database's firewall dropping packets from the agent's IP, which she resolves by updating the firewall rules.
+
+After a short (<20s) time the threshold is reached and the agent changes its status to `healthy`.
+
+The warning on the resource tile goes away.
+
+Alice checks the health status manually with tctl:
+
+```
+$ tctl get db_server/example
+
+kind: db_server
+metadata:
+  expires: "2025-02-21T02:03:47.398926Z"
+  name: example
+  revision: 43e96231-faaf-43c3-b9b8-15cf91813389
+spec:
+  ...*snip*...
+  host_id: 278be63c-c87e-4d7e-a286-86002c7c45c3
+  hostname: mars.internal
+  ...*snip*...
+status:
+  target_health:
+    addr: example.com:5432
+    protocol: TCP
+    transition_timestamp: "2025-02-19T01:53:39.144218Z"
```

Comment:
I think that in addition to "transition timestamp", we also need the time of the last health check.

Reply:
IMO we either slightly change how database heartbeats are done, so the auth is in position to decide how and when things get written (even more frequently than now, if it's necessary!), or you run your health checks as often as you want but only ever heartbeat with the exact same cadence as now, by updating the `CompareServers` function to ignore changes to the health status.

Please, please, PLEASE do not increase the upsert frequency or implement logic to optionally do so in the agent, it's going to be hell to reconcile with any scaling change we might want to do (or get forced to do in a rush) in the future, and it's going to be really problematic if some bug happens or some mistuned parameters cause a problem and the only fix is to get customers to update their agents in a rush.

Reply:
@GavinFrazar @espadolini FWIW this:

> or you run your health checks as often as you want but only ever heartbeat with the exact same cadence as now, by updating the CompareServers function to ignore changes to the health status.

from Edoardo's message above is exactly what I meant as well. Let's document we will implement it this way and proceed with implementation to not prolong the debate.